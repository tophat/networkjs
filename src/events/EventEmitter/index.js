import Event from '../Event'

export default class EventEmitter {
    constructor() {
        this.events = {}
    }

    registerEvent(eventName) {
        const event = new Event(eventName)
        this.events[eventName] = event
    }

    dispatchEvent(eventName, ...eventArgs) {
        this.events[eventName].callbacks.forEach(callback => {
            callback(...eventArgs)
        })
    }

    addEventListener(eventName, callback) {
        this.events[eventName].registerCallback(callback)
    }
}
