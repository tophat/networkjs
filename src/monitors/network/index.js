import { NetworkStatus } from '../../constants'

class NetworkMonitor {
    constructor(emitter) {
        this.emitter = emitter
        this.emitOnlineEvent = () => {
            this.emitter.dispatchEvent(NetworkStatus.ONLINE)
        }
        this.emitOfflineEvent = () => {
            this.emitter.dispatchEvent(NetworkStatus.OFFLINE)
        }
        this.initialize()
    }

    initialize() {
        window.addEventListener('online', this.emitOnlineEvent)
        window.addEventListener('offline', this.emitOfflineEvent)
    }

    pause() {
        window.removeEventListener('online', this.emitOnlineEvent)
        window.removeEventListener('offline', this.emitOfflineEvent)
    }

    resume() {
        this.initialize()
    }
}

export default NetworkMonitor
