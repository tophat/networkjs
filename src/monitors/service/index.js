import { NetworkStatus } from '../../constants'
import { ServiceDefaults } from './Service.constants'

class ServiceMonitor {
    constructor(
        emitter,
        {
            prefixes = ServiceDefaults.PREFIXES,
            statuses = ServiceDefaults.STATUSES,
            failureThreshold = ServiceDefaults.FAILURE_THRESHOLD,
            decrementTime = ServiceDefaults.DECREMENT_TIME,
        },
    ) {
        this.emitter = emitter
        this.prefixes = prefixes
        if (
            Array.isArray(prefixes) &&
            prefixes.every(p => typeof p === 'string')
        ) {
            this.isPrefixArray = true
            this.regexes = prefixes.map(p => new RegExp(p))
        } else if (prefixes === '*') {
            this.isPrefixArray = false
        } else {
            throw new Error(`prefixes must be an array or '*'`)
        }
        this.statuses = statuses

        if (typeof failureThreshold === 'number') {
            this.failureThreshold = failureThreshold
        } else {
            throw new Error(`failureThreshold must be a number`)
        }
        this.decrementTime = decrementTime

        this.initialize()
    }

    initialize() {
        this.paused = false
        this.failureCounts = {}
        if (this.isPrefixArray) {
            for (let i = 0; i < this.prefixes.length; i++) {
                this.failureCounts[this.prefixes[i]] = 0
            }
        } else {
            this.failureCount = 0
        }
    }

    getPrefixForUrl(url) {
        if (!this.isPrefixArray) return undefined

        for (let i = 0; i < this.regexes.length; i++) {
            if (this.regexes[i].test(url)) {
                return this.prefixes[i]
            }
        }
        return undefined
    }

    getCountForPrefix(prefix) {
        return this.isPrefixArray
            ? this.failureCounts[prefix]
            : this.failureCount
    }

    increment(prefix) {
        if (this.isPrefixArray) {
            this.failureCounts[prefix]++
        } else {
            this.failureCount++
        }
    }

    decrement(prefix) {
        if (this.isPrefixArray) {
            this.failureCounts[prefix]--
        } else {
            this.failureCount--
        }
    }

    waitAndDecrement(prefix) {
        setTimeout(() => {
            if (!this.paused) {
                this.decrement(prefix)
                const count = this.getCountForPrefix(prefix)
                if (count === this.failureThreshold - 1) {
                    this.emitter.dispatchEvent(NetworkStatus.RESOLVED, prefix)
                }
            }
        }, this.decrementTime)
    }

    handleError(url, status) {
        if (!this.paused && this.statuses.includes(status)) {
            const prefix = this.getPrefixForUrl(url)
            this.increment(prefix)
            const count = this.getCountForPrefix(prefix)
            if (count === this.failureThreshold) {
                this.emitter.dispatchEvent(NetworkStatus.DEGRADED, prefix)
            }
            this.waitAndDecrement(prefix)
        }
    }

    pause() {
        this.paused = true
    }

    resume() {
        this.initialize()
    }
}

export default ServiceMonitor
