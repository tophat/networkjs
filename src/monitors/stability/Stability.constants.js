export const StabilityDefaults = {
    MAX_BUFFER_SIZE: 10,
    SPEED_THRESHOLD: 100,
}
