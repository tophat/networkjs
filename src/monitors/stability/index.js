import { NetworkStatus } from '../../constants'
import { StabilityDefaults } from './Stability.constants'

export const delay = async interval => {
    return await new Promise(done => setTimeout(() => done(), interval))
}

export const ping = async resource => {
    try {
        await fetch(resource)
    } catch (e) {
        return false
    }
    return true
}

class StabilityMonitor {
    constructor(
        emitter,
        {
            resource,
            interval = StabilityDefaults.INTERVAL,
            requestThreshold = StabilityDefaults.REQUEST_THRESHOLD,
            durationThreshold = StabilityDefaults.DURATION_THRESHOLD,
        },
    ) {
        this.emitter = emitter
        this.resource = resource
        this.interval = interval
        this.requestThreshold = requestThreshold
        this.durationThreshold = durationThreshold

        this.initialize()
    }

    initialize() {
        this.consecutiveSlowRequestCount = 0
        this.paused = false
        this.run()
    }

    pause() {
        this.paused = true
    }

    resume() {
        this.initialize()
    }

    async run() {
        let shouldDelay = true
        const start = window.performance.now()
        const success = await ping(this.resource)
        if (success) {
            if (window.performance.now() - start > this.durationThreshold) {
                this.consecutiveSlowRequestCount++
                if (this.consecutiveSlowRequestCount < this.requestThreshold) {
                    shouldDelay = false
                } else if (
                    this.consecutiveSlowRequestCount ===
                        this.requestThreshold &&
                    !this.paused
                ) {
                    this.emitter.dispatchEvent(NetworkStatus.UNSTABLE)
                }
            } else {
                if (
                    this.consecutiveSlowRequestCount >= this.requestThreshold &&
                    !this.paused
                ) {
                    this.emitter.dispatchEvent(NetworkStatus.STABLE)
                }
                this.consecutiveSlowRequestCount = 0
            }
        }
        shouldDelay && (await delay(this.interval))
        !this.paused && (await this.run())
    }
}

export default StabilityMonitor
