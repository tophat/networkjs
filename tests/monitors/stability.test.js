import * as stability from '../../src/monitors/stability'
import EventEmitter from '../../src/events/EventEmitter'

const delay = stability.delay
const ping = stability.ping
const StabilityMonitor = stability.default

describe('Stability Monitor', () => {
    describe('delay', () => {
        it('waits the provided interval', () => {
            jest.useFakeTimers()
            delay(1000)

            expect(setTimeout).toHaveBeenCalledTimes(1)
            expect(setTimeout).toHaveBeenLastCalledWith(
                expect.any(Function),
                1000,
            )
        })
    })

    describe('ping', () => {
        it('returns true on fetch success', async () => {
            global.fetch = jest
                .fn()
                .mockImplementation(() =>
                    Promise.resolve({ ok: true, Id: '123' }),
                )
            const result = await ping('TEST')

            expect(global.fetch).toHaveBeenCalledTimes(1)
            expect(result).toBe(true)
        })

        it('returns false on fetch error', async () => {
            global.fetch = jest.fn().mockImplementation(() => Promise.reject())
            const result = await ping('TEST')

            expect(global.fetch).toHaveBeenCalledTimes(1)
            expect(result).toBe(false)
        })
    })

    describe('StabilityMonitor', () => {
        let monitor
        const emitter = new EventEmitter()

        beforeEach(() => {
            monitor = new StabilityMonitor(emitter, {
                resource: 'URL',
                interval: 5000,
                requestThreshold: 2000,
                durationThreshold: 2,
            })
        })

        describe('constructor', () => {
            it('initializes with the correct props', () => {
                expect(monitor).toMatchSnapshot()
            })
        })

        describe('initialize', () => {
            it('sets some default properties and runs monitor', () => {
                const runSpy = (StabilityMonitor.prototype.run = jest.fn())
                monitor.initialize()

                expect(monitor.consecutiveSlowRequestCount).toBe(0)
                expect(monitor.paused).toBe(false)
                expect(runSpy).toHaveBeenCalledTimes(1)
            })
        })

        describe('pause', () => {
            it('sets paused to true', () => {
                monitor.pause()

                expect(monitor.paused).toBe(true)
            })
        })

        describe('resume', () => {
            it('re-initializes monitor', () => {
                const initializeSpy = (StabilityMonitor.prototype.initialize = jest.fn())
                monitor.resume()

                expect(monitor.consecutiveSlowRequestCount).toBe(0)
                expect(monitor.paused).toBe(false)
                expect(initializeSpy).toHaveBeenCalledTimes(1)
            })
        })

        // describe('run', () => {
        //     it('pings the provided resource', async () => {
        //         const pingSpy = jest
        //             .spyOn(stability, 'ping')
        //             .mockImplementation(async () => false)
        //         await monitor.run()

        //         expect(pingSpy).toHaveBeenCalledTimes(1)
        //         expect(pingSpy).toHaveBeenLastCalledWith('URL')
        //     })
        // })
    })
})

/*
class StabilityMonitor {

    async run() {
        let shouldDelay = true
        const start = window.performance.now()
        const success = await ping(this.resource)
        if (success) {
            if (window.performance.now() - start > this.durationThreshold) {
                this.consecutiveSlowRequestCount++
                if (this.consecutiveSlowRequestCount < this.requestThreshold) {
                    shouldDelay = false
                } else if (
                    this.consecutiveSlowRequestCount ===
                        this.requestThreshold &&
                    !this.paused
                ) {
                    this.emitter.dispatchEvent(NetworkStatus.UNSTABLE)
                }
            } else {
                if (
                    this.consecutiveSlowRequestCount >= this.requestThreshold &&
                    !this.paused
                ) {
                    this.emitter.dispatchEvent(NetworkStatus.STABLE)
                }
                this.consecutiveSlowRequestCount = 0
            }
        }
        shouldDelay && (await delay(this.interval))
        !this.paused && (await this.run())
    }
}

export default StabilityMonitor
*/
