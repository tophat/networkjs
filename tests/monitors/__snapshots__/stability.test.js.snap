// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stability Monitor StabilityMonitor constructor initializes with the correct props 1`] = `
StabilityMonitor {
  "consecutiveSlowRequestCount": 0,
  "durationThreshold": 2,
  "emitter": EventEmitter {
    "events": Object {},
  },
  "interval": 5000,
  "paused": false,
  "requestThreshold": 2000,
  "resource": "URL",
}
`;
